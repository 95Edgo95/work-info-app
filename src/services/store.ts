import createSagaMiddleware, {SagaMiddleware} from "redux-saga";
import {applyMiddleware, compose, createStore} from "redux";
import reducer from "services/reducer";
import saga from "services/saga";
import {fromJS} from "immutable";

let store: any;

const initialState: any = fromJS({});

const sagaMiddleware: SagaMiddleware<any> = createSagaMiddleware();

if (process.env && process.env.NODE_ENV !== "production") {
    const composeEnhancers: any = (window as any).__REDUX_DEVTOOLS_EXTENSION_COMPOSE__ || compose;
    store = createStore(reducer, initialState, composeEnhancers(applyMiddleware(sagaMiddleware)));
} else {
    store = createStore(reducer, initialState, compose(applyMiddleware(sagaMiddleware)));
}

sagaMiddleware.run(saga);

export default store;
