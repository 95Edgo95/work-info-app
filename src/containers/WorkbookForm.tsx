import {attemptCreateWorkbook, attemptGetWorkbook, attemptUpdateWorkbook} from "store/workbook/WorkbookActionCreators";
import {mapDispatchToProps, mapStateToProps} from "helpers/ReduxHelper";
import {IWorkbook} from "store/workbook/WorkbookReducer";
import {ISelectorVariables} from "services/selector";
import {RouteComponentProps} from "react-router";
import {connect} from "react-redux";
import Form from "components/Form";
import * as React from "react";
import {Map} from "immutable";

const inputs: any[] = [
  {
    placeholder: "First Name",
    name: "firstName",
    defaultValue: "",
  },
  {
    placeholder: "Last Name",
    defaultValue: "",
    name: "lastName",
  },
  {
    placeholder: "Passport",
    defaultValue: "",
    name: "passport",
  },
  {
    placeholder: "Email",
    defaultValue: "",
    name: "email",
  },
  {
    placeholder: "Birth Date",
    name: "birthDate",
    defaultValue: "",
    type: "date"
  },
];

interface IWorkbookFormReduxProps {
  workbooks: Map<string, IWorkbook>;
}

interface IWorkbookFormDispatchProps {
  attemptUpdateWorkbook: (data: any, id: string) => void;
  attemptCreateWorkbook: (data: any) => void;
  attemptGetWorkbook: (id: string) => void;
}

interface IWorkbookFormProps extends RouteComponentProps<{ id: string }>, IWorkbookFormReduxProps, IWorkbookFormDispatchProps {}

function WorkbookForm(props: IWorkbookFormProps) {
  const {
    attemptCreateWorkbook,
    attemptUpdateWorkbook,
    match: {params: {id}},
    attemptGetWorkbook,
    workbooks,
    history,
  } = props;

  const workbook: any = id ? workbooks.get(id) : Map({
    firstName: "",
    lastName: "",
    passport: "",
    email: "",
    birthDate: "",
  });

  const validations = {
    firstName: [
      {
        type: "required",
        message: "Required!",
      },
      {
        type: "regex",
        option: /^[a-zA-Z]+$/gm,
        message: "Only letters.",
      }
    ],
    lastName: [
      {
        type: "required",
        message: "Required!",
      },
      {
        type: "regex",
        option: /^[a-zA-Z]+$/gm,
        message: "Only letters.",
      }
    ],
    passport: [
      {
        type: "required",
        message: "Required!",
      },
      {
        type: "regex",
        option: /^[a-zA-Z0-9]+$/g,
        message: "Invalid passport number!",
      }
    ],
    email: [
      {
        type: "required",
        message: "Required!",
      },
      {
        type: "regex",
        option: /^(([^<>()\[\]\\.,;:\s@"]+(\.[^<>()\[\]\\.,;:\s@"]+)*)|(".+"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/,
        message: "Invalid email!",
      }
    ],
    birthDate: [
      {
        type: "required",
        message: "Required!",
      },
      {
        type: "regex",
        option: /^[0-9]{1,2}[\/\s.,-][0-9]{1,2}[\/\s.,-][0-9]{4}$/gm,
        message: "Invalid Date!",
      },
      {
        type: "is18+",
        message: "Must be 18+!",
      }
    ],
  };

  React.useEffect(() => {
    if (id && !workbook) {
      attemptGetWorkbook(id);
    }
  }, []);

  if (!workbook) {
    return null;
  }

  return (
    <Form
      attemptUpdateEntity={attemptUpdateWorkbook}
      attemptCreateEntity={attemptCreateWorkbook}
      entitiesSize={workbooks.size}
      validations={validations}
      entity={workbook}
      history={history}
      inputs={inputs}
    />
  );
}

const actions = {
  attemptCreateWorkbook,
  attemptUpdateWorkbook,
  attemptGetWorkbook,
};

const selectorVariables: ISelectorVariables = {
  workbook: {
    workbooks: true,
  }
};

export default connect<IWorkbookFormReduxProps, IWorkbookFormDispatchProps, any>(mapStateToProps(selectorVariables), mapDispatchToProps(actions))(WorkbookForm);
