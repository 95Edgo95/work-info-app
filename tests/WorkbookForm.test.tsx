import WorkbookForm from "../src/containers/WorkbookForm";
import {BrowserRouter} from "react-router-dom";
import store from "../__mocks__/storeMock";
import {Provider} from "react-redux";
import { mount } from "enzyme";
import * as React from "react";

describe('WorkbookForm', () => {
  it('should render correctly', () => {
    const component = mount(
      <Provider store={store}>
        <BrowserRouter>
          <WorkbookForm match={{ params: { id: "asdf" } }} />
        </BrowserRouter>
      </Provider>
    );

    expect(component).toMatchSnapshot();
    component.unmount();
  });
});
