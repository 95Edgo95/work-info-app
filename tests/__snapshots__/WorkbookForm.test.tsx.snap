// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WorkbookForm should render correctly 1`] = `
<Provider
  store={
    Object {
      "dispatch": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
      Symbol(observable): [Function],
    }
  }
>
  <BrowserRouter>
    <Router
      history={
        Object {
          "action": "POP",
          "block": [Function],
          "createHref": [Function],
          "go": [Function],
          "goBack": [Function],
          "goForward": [Function],
          "length": 1,
          "listen": [Function],
          "location": Object {
            "hash": "",
            "pathname": "/",
            "search": "",
            "state": undefined,
          },
          "push": [Function],
          "replace": [Function],
        }
      }
    >
      <Connect(WorkbookForm)
        match={
          Object {
            "params": Object {
              "id": "asdf",
            },
          }
        }
      >
        <WorkbookForm
          attemptCreateWorkbook={[Function]}
          attemptGetWorkbook={[Function]}
          attemptUpdateWorkbook={[Function]}
          match={
            Object {
              "params": Object {
                "id": "asdf",
              },
            }
          }
          workbooks={Immutable.Map {}}
        />
      </Connect(WorkbookForm)>
    </Router>
  </BrowserRouter>
</Provider>
`;
